plugin/restore/mysql/mysql.go

```diff
@@ -457,87 +484,144 @@ func getPITROldDatabaseName(database string, suffixTs int64) string {
 	return getSafeName(database, suffix)
 }
 
-// FetchArchivedBinlogFiles downloads the missing binlog files from the remote instance to `binlogDir`,
-// but exclude latest binlog. We may  download the latest binlog only when doing PITR.
-func (r *Restore) FetchArchivedBinlogFiles(ctx context.Context) error {
-	binlogFilesLocal, err := ioutil.ReadDir(r.binlogDir)
+// FetchBinlogFilesToTargetTs downloads the locally missing binlog files from the remote instance to `binlogDir`.
+// After downloading a binlog file, we check it's first event timestamp. If the timestamp is larger than
+// targetTs, we stop the following downloads and return eagerly, because now the local binlog files contain
+// all the binlog events we need for this recovery task.
+func (r *Restore) FetchBinlogFilesToTargetTs(ctx context.Context, targetTs int64) error {
+	binlogFilesInfoLocal, err := ioutil.ReadDir(r.binlogDir)
 	if err != nil {
 		return err
 	}
-
-	binlogFilesOnServer, err := r.getBinlogFilesMetaOnServer(ctx)
-	if err != nil {
-		return err
+	var binlogFilesLocal []BinlogFile
+	for _, fileInfo := range binlogFilesInfoLocal {
+		binlogFile, err := newBinlogFile(fileInfo.Name(), fileInfo.Size())
+		if err != nil {
+			return err
+		}
+		binlogFilesLocal = append(binlogFilesLocal, binlogFile)
 	}
 
-	latestBinlogFileOnServer, err := r.getLatestBinlogFileMeta(ctx)
+	binlogFilesOnServerSorted, err := r.GetSortedBinlogFilesMetaOnServer(ctx)
 	if err != nil {
 		return err
 	}
-
-	// build a local file size map from file name to size
-	localFileMap := make(map[string]int64)
-
-	for _, localFile := range binlogFilesLocal {
-		localFileMap[localFile.Name()] = localFile.Size()
+	if len(binlogFilesOnServerSorted) == 0 {
```

Comment:
Also checking local binlogs could be put before fetching binlogs from the remote server, since if the local binlog files are enough, there is no need to fetch from the remote server.

Reply:
@tianzhou Validating local binlog files' size requires fetching the binlog metadata from server.
The downloading is postponed and prevented as possible.